src/Storages/MaterializedView/RefreshTask.cpp

```diff
@@ -424,37 +435,56 @@ void RefreshTask::executeRefreshUnlocked(std::shared_ptr<StorageMaterializedView
         }
 
         /// Exchange tables.
-        stale_table = view->exchangeTargetTable(refresh_query->table_id, refresh_context);
+        if (!append)
+            table_to_drop = view->exchangeTargetTable(refresh_query->table_id, refresh_context);
     }
     catch (...)
     {
-        try
-        {
-            InterpreterDropQuery::executeDropQuery(
-                ASTDropQuery::Kind::Drop, view->getContext(), refresh_context, refresh_query->table_id, /*sync*/ false, /*ignore_sync_setting*/ true);
-        }
-        catch (...)
-        {
-            tryLogCurrentException(log, "Failed to drop temporary table after a failed refresh");
-            /// Let's ignore this and keep going, at risk of accumulating many trash tables if this keeps happening.
-        }
+        if (table_to_drop.has_value())
+            view->dropTempTable(table_to_drop.value(), refresh_context);
         throw;
     }
 
     /// Drop the old table (outside the try-catch so we don't try to drop the other table if this fails).
-    InterpreterDropQuery::executeDropQuery(ASTDropQuery::Kind::Drop, view->getContext(), refresh_context, stale_table, /*sync*/ true, /*ignore_sync_setting*/ true);
+    if (table_to_drop.has_value())
+        view->dropTempTable(table_to_drop.value(), refresh_context);
```

Comment:
Removing outdated comment, leaving the inner try-catch, the idea is that it's better to consider the refresh successful in this case to avoid retries.